feat: permissions impl and demo

DmitryEfimenko opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

-   [X] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.4/)
-   [X] Tests for the changes have been added (for bug fixes/features)
-   [X] Docs have been added/updated (for bug fixes/features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

-   [ ] Bugfix
-   [X] Feature
-   [ ] Refactoring (no functional changes, no API changes)
-   [ ] Other... Please describe:

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying or link to a relevant issue. -->

Issue Number: N/A

## What is the new behavior?
Initial implementation of PermissionsService

## Does this PR introduce a breaking change?

-   [ ] Yes
-   [X] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

* I am not sure what process you usually follow for publishing demo to gh-pages. I haven't found anything about it in other ng-web-api repos. I recommend using package [gh-pages](https://www.npmjs.com/package/gh-pages). It's super easy.
* The demo project does not start in ssr mode due to an error in ngx-highlightjs. Tracking it down further it looks like it's due to the use of `animationFrameScheduler` from rxjs. I don't think it's super important for our use case since we're publishing demo to gh-pages without the use of ssr, but I [opened an issue with rxjs](https://github.com/ReactiveX/rxjs/issues/5940) anyway.

DmitryEfimenko committed:
feat: permissions impl and demo

DmitryEfimenko committed:
fix(build): include lib in the build

DmitryEfimenko committed:
fix(demo): add specs

DmitryEfimenko commented:
I've fixed a few issues that I thought are causing Travis CI to fail, but now I'm out of ideas about how to fix the error it throws. I'm on Windows and the posttest command does not work for me locally so I'm a bit limited there. The error:
> \> node scripts/coveralls.js
/home/travis/build/ng-web-apis/permissions/node_modules/coveralls/bin/coveralls.js:19
      throw err;
      ^
Bad response: 422 {"message":"Couldn't find a repository matching this job.","error":true}
child_process.js:669
    throw err;
    ^
Error: Command failed: cat coverage/demo/lcov.info | coveralls

IKatsuba commented:
@DmitryEfimenko, could you set up CI/CD using Github Actions like in this repo https://github.com/ng-web-apis/workers? We deliberately abandoned Travis CI

IKatsuba commented on projects/demo/src/app/app.component.ts:
```diff
@@ -8,6 +11,26 @@ import {PermissionsService} from '@ng-web-apis/permissions';
+                    'After you acknowledge this message the browser will ask your permission to allow location access!',
+                );
+                // calling getCurrentPosition() results in a browser asking for permission if the current status is "prompt"
+                navigator.geolocation.getCurrentPosition(() => {});
```
Maybe use the NAVIGATOR token  here? https://github.com/ng-web-apis/common/blob/master/projects/common/src/tokens/navigator.ts

DmitryEfimenko replied:
sounds good. Will update

IKatsuba commented on projects/demo/src/app/samples/basic.ts:
```diff
@@ -0,0 +1,17 @@
+  geolocationState$ = this.permissionsService.state('geolocation');
+
+  constructor(
+    @Inject(PermissionsService) readonly permissionsService: PermissionsService,
```
Why is Inject used here?

DmitryEfimenko replied:
good catch. Will update.

IKatsuba commented on projects/permissions/src/services/permissions.service.ts:
```diff
@@ -1,6 +1,65 @@
+                    res => {
+                        permissionStatus = res;
+                        subscriber.next(res.state);
+                        permissionStatus.addEventListener('change', onChange);
```
You can use the fromEvent factory and the switchMap operator, for example. then you won't need to call the removeEventListener function

DmitryEfimenko replied:
Elegant! Will update

DmitryEfimenko committed:
chore: switch to GitHub Actions

DmitryEfimenko committed:
chore: address PR comments and remove Travis

DmitryEfimenko commented:
I've removed Travis and added Github Actions. They won't kick in for this PR though. Also, some secrets might need to be set up by the repo owner.
I also addressed @IKatsuba's comments. Please have a look.

DmitryEfimenko committed:
chore(docs): update sample

IKatsuba commented on projects/permissions/src/services/permissions.service.ts:
```diff
@@ -1,6 +1,51 @@
+
+            const sub = from(this.permissions.query(descriptor))
+                .pipe(
+                    tap(permissionStatus => subscriber.next(permissionStatus.state)),
```
IMHO, that will be more readable (the tap operator removed)
```
switchMap(permissionStatus => fromEvent(permissionStatus, 'change').pipe(startWith(permissionStatus.state))),
```

DmitryEfimenko replied:
I don't know. We're getting into multiple pipe nesting. Besides, in order to make that work, the code should be something like this:
```ts
const sub = from(this.permissions.query(descriptor))
    .pipe(
      switchMap(permissionStatus => fromEvent(permissionStatus, 'change').pipe(
        map(ev => (ev.target as PermissionStatus).state),
        startWith(permissionStatus.state),
      )),
    )
    .subscribe(
        state => subscriber.next(state),
        error => subscriber.error(error),
    );
```
I'd argue this is less readable.

IKatsuba replied:
Sorry, I meant
```typescript
switchMap(permissionStatus => fromEvent(permissionStatus, 'change').pipe(startWith(permissionStatus))),
```

DmitryEfimenko replied:
that's basically what I have in my example there. Except that I can't use the `startWith` as a first operator in the inner pipe because `fromEvent(permissionStatus, 'change')` returns an object of type `Event`. So I can't "start it with" an object of type `PermissionState`. There needs to be a `map` in between to match types.

Unless I misunderstand something. Please provide a full snippet.

DmitryEfimenko replied:
oh, I see the difference. Still though there needs to be a `map`:
```ts
const sub = from(this.permissions.query(descriptor))
  .pipe(
    switchMap(permissionStatus => fromEvent(permissionStatus, 'change').pipe(
      map(ev => (ev.target as PermissionStatus)),
      startWith(permissionStatus),
    )),
  )
  .subscribe(
      permissionStatus => subscriber.next(permissionStatus.state),
      error => subscriber.error(error),
  );
```

IKatsuba replied:
You're right. I read inattentively

waterplea commented on projects/permissions/src/services/permissions.service.ts:
```diff
@@ -1,6 +1,51 @@
+                return;
+            }
+
+            const sub = from(this.permissions.query(descriptor))
```
How about something like that?

```ts
    return from(
        this.permissions.query(descriptor).pipe(
            switchMap(status =>
                fromEvent(status, 'change').pipe(
                    startWith(null),
                    map(() => status.state),
                ),
            ),
        ),
    ).subscribe(subscriber);
```

`PermissionState` is live so we can use `map(() => status.state)` to return actual state without typecasting event target. We can also just pass `subscriber`  to `subscribe` this way and return resulting `Subscription` as `TeardownLogic`.

DmitryEfimenko replied:
sneaky! Good thinking. I'll change to that.

DmitryEfimenko committed:
chore(core): refactor rxjs

DmitryEfimenko commented:
@MarsiBarsi please have a look when you have a chance

MarsiBarsi reviewed (approved):
I think it is ready to release 🎉 
Great job!

DmitryEfimenko commented:
@waterplea it's all in your hands now :)
Please let me know if you see any other areas where we can improve

waterplea commented on projects/permissions/src/tokens/permissions.ts:
```diff
@@ -0,0 +1,9 @@
+import {inject, InjectionToken} from '@angular/core';
+import {NAVIGATOR} from '@ng-web-apis/common';
+
+export const PERMISSIONS = new InjectionToken<Permissions>(
```
I wonder if we should add `null` as possible type for unsupporting browsers, or is it fine as is, because we have SUPPORT token. What do you think, @MarsiBarsi ?

waterplea commented:
Didn't really have time to look through demo and tests, the actual code looks good to me. Left one more comment to think about. @MarsiBarsi will you assist with publishing? I will be unavailable for a week or so.

DmitryEfimenko commented:
hey guys, how's it going? Still celebrating the new year? :D

MarsiBarsi commented:
Dmitry, I am sorry :)
I missed the message 🙈 Let me merge and finish it today
